pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -477,6 +477,15 @@ func containerSucceeded(c *v1.Container, podStatus *kubecontainer.PodStatus) boo
 func (m *kubeGenericRuntimeManager) computePodActions(pod *v1.Pod, podStatus *kubecontainer.PodStatus) podActions {
 	klog.V(5).Infof("Syncing Pod %q: %+v", format.Pod(pod), pod)
 
+	// All containers in the pod are done and the pod may be cleaned up
+	if pod.Status.Phase == v1.PodSucceeded || pod.Status.Phase == v1.PodFailed {
```

Comment:
This seems like it is addressing a bug in the code below, which should handle this situation.

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L496

Can you include this logic there instead of placing it separately?